src/watch/index.ts

```diff
@@ -243,31 +224,43 @@ export class Task {
 					result
 				});
 			})
-			.catch((error: Error) => {
+			.catch((error: RollupError) => {
 				if (this.closed) return;
 
-				if (this.cache) {
-					// this is necessary to ensure that any 'renamed' files
-					// continue to be watched following an error
-					if (this.cache.modules) {
-						this.cache.modules.forEach(module => {
-							if (module.transformDependencies) {
-								module.transformDependencies.forEach(depId => {
-									this.watchFile(depId, true);
-								});
-							}
-						});
-					}
-					this.watchFiles.forEach(id => {
+				if (Array.isArray(error.watchFiles)) {
+					for (const id of error.watchFiles) {
```

Comment:
As I see it, none of the other logic is needed if we only remove files on successful builds. I also removed the `absolutePath` logic as this is already caught in the file watcher itself; otherwise, the logic between successful runs and errors would be asymmetric. If this actually caused issues for you, we should add a test.

Reply:
I believe this will still happen. At the moment, those will be filtered out here: https://github.com/rollup/rollup/blob/1429d571908319df6c1b9c8f722c525c2077c66e/src/watch/fileWatchers.ts#L52

But we can save us a few unneeded `fs.statSync`. I think the proper place would be to do this in the watcher logic—similar how you did it but symmetric for all situations—as Rollup core should not know about which files are "watchable". In the future, I hope we can make the watcher completely pluggable so that you can e.g. use TypeScripts watcher, but that's another story. I will add this before we merge this.

Reply:
Ah, it seems that just checking for `isAbsolute` is not enough because it is supported (and tested) to add non-absolute watch files via `this.addWatchFile`. So I guess we stick with the old logic